packages/firebase_auth/firebase_auth_platform_interface/lib/src/types.dart

```diff
@@ -0,0 +1,242 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'package:meta/meta.dart';
+
+typedef void PhoneVerificationCompleted(AuthCredential phoneAuthCredential);
+typedef void PhoneVerificationFailed(AuthException error);
+typedef void PhoneCodeSent(String verificationId, [int forceResendingToken]);
+typedef void PhoneCodeAutoRetrievalTimeout(String verificationId);
+
+/// Represents a `User` in Firebase.
+///
+/// See also: https://firebase.google.com/docs/reference/js/firebase.User
+class PlatformUser {
+  const PlatformUser({
+    @required this.providerId,
+    @required this.uid,
+    this.displayName,
+    this.photoUrl,
+    this.email,
+    this.phoneNumber,
+    this.creationTimestamp,
+    this.lastSignInTimestamp,
+    @required this.isAnonymous,
+    @required this.isEmailVerified,
+    @required this.providerData,
+  });
+
+  final String providerId;
+  final String uid;
+  final String displayName;
+  final String photoUrl;
+  final String email;
+  final String phoneNumber;
+  final int creationTimestamp;
+  final int lastSignInTimestamp;
+  final bool isAnonymous;
+  final bool isEmailVerified;
+  final List<PlatformUserInfo> providerData;
+}
+
+/// Represents a `UserInfo` from Firebase.
+///
+/// See also: https://firebase.google.com/docs/reference/js/firebase.UserInfo
+class PlatformUserInfo {
+  const PlatformUserInfo({
+    @required this.providerId,
+    @required this.uid,
+    this.displayName,
+    this.photoUrl,
+    this.email,
+    this.phoneNumber,
+  });
+
+  final String providerId;
+  final String uid;
+  final String displayName;
+  final String photoUrl;
+  final String email;
+  final String phoneNumber;
+}
+
+/// Represents `AdditionalUserInfo` from Firebase.
+///
+/// See also: https://firebase.google.com/docs/reference/js/firebase.auth.html#additionaluserinfo
+class PlatformAdditionalUserInfo {
+  const PlatformAdditionalUserInfo({
+    @required this.isNewUser,
+    @required this.providerId,
+    @required this.username,
+    @required this.profile,
+  });
+
+  final bool isNewUser;
+  final String providerId;
+  final String username;
+  final Map<String, dynamic> profile;
+}
+
+class PlatformAuthResult {
+  const PlatformAuthResult({
+    @required this.user,
+    @required this.additionalUserInfo,
+  });
+
+  final PlatformUser user;
+  final PlatformAdditionalUserInfo additionalUserInfo;
+}
+
+/// Represents `AuthCredential` from Firebase.
+abstract class AuthCredential {
+  const AuthCredential(this.providerId);
+
+  /// An id that identifies the specific type of provider.
+  final String providerId;
+
+  /// Returns the data for this credential as a map.
+  Map<String, String> asMap();
+}
```

Comment:
A toString method that returns `this.asMap().toString()` may be handy for logging!

Reply:
Agreed